### ISSUE TITLE
UNUSE: doesn't work in Listener

### ISSUE THREAD

**Issue by alex-ilin** [factor/factor#2374]
1. Try the following:
```
USE: peg.parsers
integer-parser
```
Stack:
```
T{ parser f ~action-parser~ f 1260 }
```

2. Then do this:
```
USE: parser-combinators.simple
integer-parser
```
Stack:
```
T{ parser f ~action-parser~ f 1260 }
T{ apply-parser f ~promise~ ~quotation~ }
```

3. Finally, do this:
```
UNUSE: parser-combinators.simple
integer-parser
```
Stack:
```
T{ parser f ~action-parser~ f 1260 }
T{ apply-parser f ~promise~ ~quotation~ }
T{ apply-parser f ~promise~ ~quotation~ }
```

Expected stack:
```
T{ parser f ~action-parser~ f 1260 }
T{ apply-parser f ~promise~ ~quotation~ }
T{ parser f ~action-parser~ f 1260 }
```

**Comment by mrjbq7** [comment]
The restart that was triggered on the second use of ``integer-parser`` which asked you to disambiguate the one you wanted to use resulted in a ``FROM: parser.combinators.simple => integer-parser ;``, which is not affected by ``UNUSE: parser.combinators.simple``.

**Comment by AlexIljin** [comment]
Thanks for the explanation! This works for step 3:
```
FORGET: integer-parser
integer-parser
```
Stack:
```
T{ parser f ~action-parser~ f 1260 }
T{ apply-parser f ~promise~ ~quotation~ }
T{ parser f ~action-parser~ f 1260 }
```

**Comment by mrjbq7** [comment]
One could argue that ``UNUSE:`` should maybe affect ``FROM:`` in this case... what do you think?

**Comment by AlexIljin** [comment]
I'd say it's more intuitive (according to my intuition, anyway) to have UNUSE: remove all words of the vocab, as if it was never loaded (USE:d) at all. Especially since I didn't explicitly use FROM:, and that's an implementation detail that was hidden from me until your explanation.

And double-especially if there is no way to `UNFROM:` a word.

**Comment by mrjbq7** [comment]
Want to try this patch out locally and see what you think?

```diff
diff --git a/core/vocabs/parser/parser.factor b/core/vocabs/parser/parser.factor
index 7993e10c21..5076b2ce20 100644
--- a/core/vocabs/parser/parser.factor
+++ b/core/vocabs/parser/parser.factor
@@ -131,7 +131,11 @@ ERROR: unbalanced-private-declaration vocab ;
         manifest get
         [ [ load-vocab ] dip search-vocabs>> remove-eq! drop ]
         [ [ vocab-name ] dip search-vocab-names>> delete ]
-        2bi
+        [
+            [ vocab-name ] dip qualified-vocabs>> [
+                dup vocab? [ vocab>> ] unless vocab-name =
+            ] with reject! drop
+        ] 2tri
     ] [ drop ] if ;
 
 TUPLE: qualified vocab prefix words ;
```

**Comment by AlexIljin** [comment]
Works like a charm!

**Comment by AlexIljin** [comment]
Also works. Ship it!

**Comment by mrjbq7** [comment]
Fixed in 363d5684398ca37c6d99aa461d566e8facb21787.
